### ISSUE TITLE
RT-AX92U: usb0 interface is added to asus bridge but raspberry pi never gets an IP

### ISSUE THREAD

**Issue by joelmarty** [jacklul/asuswrt-usb-raspberry-pi#1]
Hi, I am using your project to setup an usb-connected pi-hole and I am having issues with the setup:
- How can I check if I need the `FAKE_ASUS_OPTWARE` variable or not ? I guess I do since setting it to true helped me to reach a stage where I can see the usb0 interface, but is there a definitive way to tell ? This is more of a feedback on documentation.
- At the moment I am at a stage where I think the setup is incomplete: the router shows an usb0 interface when running `ifconfig` but the raspberry pi does not: it only shows the wlan0, lo and since I am planning to use docker to run pihole, the docker0 interface. Here is a log of the script on the pi:
```
-- Boot d004effabab943fea4a05394edf99e67 --
Jan 14 18:19:03 pihole systemd[1]: Starting asuswrt-usb-network.service - Enable USB network when connected to Asus router>
Jan 14 18:19:05 pihole asuswrt-usb-network[340]: Setting up gadget "usbnet" with function "mass_storage"...
Jan 14 18:19:05 pihole asuswrt-usb-network[340]: Creating image file "/tmp/asuswrt-usb-network.img" (ext2, 1M)...
Jan 14 18:19:06 pihole asuswrt-usb-network[340]: Creating fake Asus Optware installation...
Jan 14 18:19:15 pihole asuswrt-usb-network[439]: Timed out for waiting the udev queue being empty.
Jan 14 18:19:16 pihole asuswrt-usb-network[340]: Waiting for the router to write mark to the image (timeout: 120s)....
Jan 14 18:19:25 pihole asuswrt-usb-network[340]: Taking down gadget "usbnet"...
Jan 14 18:19:25 pihole asuswrt-usb-network[340]: Setting up gadget "usbnet" with function "ecm"...
Jan 14 18:19:31 pihole asuswrt-usb-network[687]: Timed out for waiting the udev queue being empty.
Jan 14 18:19:31 pihole asuswrt-usb-network[340]: Checking if router is reachable (timeout: 120s)...
Jan 14 18:22:41 pihole asuswrt-usb-network[340]: Completed but couldn't determine network status (timeout reached)
Jan 14 18:22:41 pihole systemd[1]: Finished asuswrt-usb-network.service - Enable USB network when connected to Asus router.
```

Here are the relevant log lines from the router:
```
May  5 07:05:23 kernel: usb 4-1: device descriptor read/64, error -110
Jan 14 18:20:07 kernel: usb 4-1: device descriptor read/64, error -110
...
Jan 14 18:20:13 kernel: scsi 0:0:0:0: Direct-Access     Linux    File-Stor Gadget 0601 PQ: 0 ANSI: 2
Jan 14 18:20:13 kernel: sd 0:0:0:0: Attached scsi generic sg0 type 0
Jan 14 18:20:13 kernel: sd 0:0:0:0: [sda] 2048 512-byte logical blocks: (1.05 MB/1.00 MiB)
Jan 14 18:20:13 kernel: sd 0:0:0:0: [sda] Write Protect is off
Jan 14 18:20:13 kernel: sd 0:0:0:0: [sda] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
Jan 14 18:20:13 kernel: sd 0:0:0:0: [sda] Attached SCSI removable disk
...
Jan 14 18:20:20 usb: USB ext2 fs at /dev/sda mounted on /tmp/mnt/sda.
Jan 14 18:20:20 asusware: umount partition /dev/sda...
Jan 14 18:20:20 disk monitor: unmount partition
Jan 14 18:20:20 hotplug: USB partition unmounted from /tmp/mnt/sda
Jan 14 18:20:20 asusware: Automatically scan partition /dev/sda...
Jan 14 18:20:20 disk monitor: scan partition
Jan 14 18:20:20 asusware: re-mount partition /dev/sda...
Jan 14 18:20:20 disk monitor: re-mount partition
Jan 14 18:20:20 usb: USB ext2 fs at /dev/sda mounted on /tmp/mnt/sda.
Jan 14 18:20:20 asusware: done.
Jan 14 18:20:20 disk monitor: done
Jan 14 18:20:21 scripts-startup.sh: Starting custom scripts (/jffs/scripts)...
Jan 14 18:20:21 scripts-startup.sh: Waiting for mount /tmp/mnt/sda to be idle...
Jan 14 18:20:21 scripts-startup.sh: Starting /jffs/scripts/hotplug-event.sh...
Jan 14 18:20:21 scripts-startup.sh: Writing mark to mount /tmp/mnt/sda...
Jan 14 18:20:21 scripts-startup.sh: Failed to unmount /tmp/mnt/sda
Jan 14 18:20:21 hotplug-event.sh: Modified hotplug configuration
Jan 14 18:20:21 scripts-startup.sh: Starting /jffs/scripts/usb-network.sh...
Jan 14 18:20:21 rc_service: hotplug 3759:notify_rc restart_nasapps
...
Jan 14 18:20:21 kernel: EXT4-fs warning (device sda): ext4_end_bio:332: I/O error -5 writing to inode 28 (offset 0 size 0 starting block 575)
Jan 14 18:20:21 kernel: Buffer I/O error on device sda, logical block 574
Jan 14 18:20:21 kernel: EXT4-fs warning (device sda): ext4_end_bio:332: I/O error -5 writing to inode 30 (offset 0 size 0 starting block 576)
Jan 14 18:20:21 kernel: Buffer I/O error on device sda, logical block 575
Jan 14 18:20:21 kernel: EXT4-fs warning (device sda): ext4_end_bio:332: I/O error -5 writing to inode 32 (offset 0 size 0 starting block 577)
Jan 14 18:20:21 kernel: Buffer I/O error on device sda, logical block 576
Jan 14 18:20:21 kernel: blk_update_request: I/O error, dev sda, sector 0
Jan 14 18:20:21 kernel: sd 0:0:0:0: [sda] Synchronizing SCSI cache
Jan 14 18:20:22 kernel: EXT4-fs error (device sda): ext4_find_entry:1461: inode #2: comm [: reading directory lblock 0
Jan 14 18:20:22 hotplug[4222]: USB partition unmounted from /tmp/mnt/sda
Jan 14 18:20:22 rc_service: hotplug 4222:notify_rc restart_nasapps
Jan 14 18:20:22 hotplug-event.sh: Running script (args: "block" "remove")
Jan 14 18:20:28 hotplug: add net usb0.
Jan 14 18:20:28 hotplug: Got net usb0, vid 0x1d6b, pid 0x104.
Jan 14 18:20:30 hotplug: set net usb0.
Jan 14 18:20:32 hotplug-event.sh: Running script (args: "net" "add")
Jan 14 18:20:32 usb-network.sh: Bringing interface usb0 up...
Jan 14 18:20:32 usb-network.sh: Added interface usb0 to bridge br0
```

Do you have any pointers to help me debug the script ?

**Comment by jacklul** [comment]
> Timed out for waiting the udev queue being empty.

This is worrying, what OS and which Pi model?

Perhaps you could modify [this line](https://github.com/jacklul/asuswrt-usb-raspberry-pi/blob/master/asuswrt-usb-network.sh#L179C33-L179C33) to produce echo when it's successful or not, perhaps this whole IF block is not getting executed.

> How can I check if I need the `FAKE_ASUS_OPTWARE` variable or not ? I guess I do since setting it to true helped me to reach a stage where I can see the usb0 interface, but is there a definitive way to tell ? This is more of a feedback on documentation.

You want this set to true only on stock firmware since there is no way to start custom scripts without hacky way.

**Comment by joelmarty** [comment]
> This is worrying, what OS and which Pi model?

This is a Pi zero W running with latest raspberry pi OS (32b). I just assumed this message came from [this line](https://github.com/jacklul/asuswrt-usb-raspberry-pi/blob/master/asuswrt-usb-network.sh#L171C5-L171C24).

> Perhaps you could modify [this line](https://github.com/jacklul/asuswrt-usb-raspberry-pi/blob/master/asuswrt-usb-network.sh#L179C33-L179C33) to produce echo when it's successful or not, perhaps this whole IF block is not getting executed.

I will try this and report back

> You want this set to true only on stock firmware since there is no way to start custom scripts without hacky way.

Gotcha. As a feedback consider adding that precision to [this paragraph](https://github.com/jacklul/asuswrt-usb-raspberry-pi/blob/master/README.md?plain=1#L66) which is kind of misleading: it lets reader think this is only necessary in some specific cases (which are not really given), not always when you run the stock firmware.

**Comment by jacklul** [comment]
> As a feedback consider adding that precision to [this paragraph](https://github.com/jacklul/asuswrt-usb-raspberry-pi/blob/master/README.md?plain=1#L66) which is kind of misleading: it lets reader think this is only necessary in some specific cases (which are not really given), not always when you run the stock firmware.

I'll see if I can rewrite this.

**Comment by jacklul** [comment]
Something that came to my mind: 
This might be systemd unit ordering issue, while the asuswt-usb-network is running there might also be another unit setting up the networking - try generating SVG boot graph with `systemd-analyze plot` to see if anything might be interfering.

Wireless is probably not it since I keep wlan0 enabled on my Zero for emergency purposes but I do have Bluetooth disabled via config.txt.
Since you mentioned Docker that might be it.

For reference this is my config.txt:
```
# Reduce GPU memory reservation (using camera requires minimum of 64 for photos and 128 for video)
gpu_mem=16

# Default LED to off and blink on activity
dtparam=act_led_activelow=off
dtparam=act_led_trigger=actpwr

# Disable audio
dtparam=audio=off

# Disable wireless
dtoverlay=disable-bt
#dtoverlay=disable-wifi

# Enable hardware watchdog
dtparam=watchdog=on

# Overclock SD card reader (defaults to 50, requires high quality card for stability)
dtparam=sd_overclock=100

# Enable gadget mode
dtoverlay=dwc2
```

PS. Running Docker on Zero 1 is awful experience.

**Comment by joelmarty** [comment]
So I followed your advice and did the following:
- Uninstalled docker
- Added `echo` lines around line 179 of the script and one for INSTANCE_NET:
```
gadget_up() {
    udevadm settle -t 5 || :
    ls /sys/class/udc > "$CONFIGFS_DEVICE_PATH/UDC"

    local INSTANCE_NET=$(find $CONFIGFS_DEVICE_PATH/functions/ -maxdepth 2 -name "ifname" | grep -o '/*[^.]*/$' || echo "")
    echo "INSTANCE_NET=$INSTANCE_NET"
    if [ -n "$INSTANCE_NET" ] && [ -f "$CONFIGFS_DEVICE_PATH/functions/$INSTANCE_NET/ifname" ]; then
        local INTERFACE="$(cat "$CONFIGFS_DEVICE_PATH/functions/$INSTANCE_NET/ifname")"

        echo "Bringing $INTERFACE up"
        ifconfig "$INTERFACE" up
        echo "'ifconfig $INTERFACE up' returned $?"
    fi
}
```

The systemd log for the asuswrt-usb-network unit did not produce those lines:
```
Jan 20 12:31:45 pihole asuswrt-usb-network[339]: Creating image file "/tmp/asuswrt-usb-network.img" (ext2, 1M)...
Jan 20 12:31:47 pihole asuswrt-usb-network[339]: Creating fake Asus Optware installation...
Jan 20 12:31:55 pihole asuswrt-usb-network[436]: Timed out for waiting the udev queue being empty.
Jan 20 12:31:55 pihole asuswrt-usb-network[339]: INSTANCE_NET=
Jan 20 12:31:55 pihole asuswrt-usb-network[339]: Waiting for the router to write mark to the image (timeout: 120s)....
Jan 20 12:32:05 pihole asuswrt-usb-network[339]: Taking down gadget "usbnet"...
Jan 20 12:32:05 pihole asuswrt-usb-network[339]: Setting up gadget "usbnet" with function "ecm"...
Jan 20 12:32:12 pihole asuswrt-usb-network[669]: Timed out for waiting the udev queue being empty.
Jan 20 12:32:12 pihole asuswrt-usb-network[339]: INSTANCE_NET=
Jan 20 12:32:13 pihole asuswrt-usb-network[339]: Checking if router is reachable (timeout: 120s)...
```
The udev error keeps happening, which seems to make the INSTANCE_NET being empty

**Comment by joelmarty** [comment]
I tried running the commands of that function one by one (after the pi has booted, not sure of the impact) on [line 174](https://github.com/jacklul/asuswrt-usb-raspberry-pi/blob/master/asuswrt-usb-network.sh#L174), `find /sys/kernel/config/usb_gadget/usbnet/functions/ -maxdepth 2 -name "ifname"` yields `/sys/kernel/config/usb_gadget/usbnet/functions/ecm.0/ifname` but after being piped into `grep -o '/*[^.]*/$'`, the result is empty, so the `|| echo ""` is executed.

It seems that for some reason that grep expression is not valid, at least for my system ?

**Comment by joelmarty** [comment]
I replaced the grep in line 174 with the following line: `local INSTANCE_NET=$(find $CONFIGFS_DEVICE_PATH/functions/ -maxdepth 2 -name "ifname" | xargs dirname | xargs basename || echo "")`

Now, $INSTANCE_NET contains the `ecm.0` string, and I reach the `ifconfig usb0 up`.
It was then a matter of making NetworkManager manage the usb0 interface to get an IP and connectivity.

I sill get `Timed out for waiting the udev queue being empty.` but it does not seem to be a problem.

I can open a PR to change the script but I am unsure if using xargs is the best option in terms of portability.

**Comment by jacklul** [comment]
> It seems that for some reason that grep expression is not valid, at least for my system ?

I just noticed it doesn't seem to work for me now either. Perhaps bash/grep update at some point broke it!

Can you check if `asuswrt-usb-network.sh` from the **develop** branch works for you? You can just copy it onto the Pi and `bash asuswrt-usb-network.sh status`, if it shows "Interface: usb0" line then it will be working.

Regarding the IP issue, I don't think I have NetworkManager on my Pi but that might be the issue. You could just make it assign a static IP to the interface.

PS. If you're annoyed by the unmount errors try this: `nvram set stop_fsck=1 && nvram commit`

**Comment by joelmarty** [comment]
I can confirm the develop branch works 👍🏻

**Comment by joelmarty** [comment]
Fixed by 40c7496
